sdcm/provision/gce/provisioner.py

```diff
@@ -0,0 +1,133 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright (c) 2022 ScyllaDB
+from __future__ import annotations
+import datetime
+import logging
+from typing import Dict, List
+
+from libcloud.compute.base import Node, NodeLocation
+
+from sdcm.keystore import KeyStore
+from sdcm.provision.gce import GCE_REGIONS
+from sdcm.provision.gce.disk_struct_provider import DiskStructProvider, DiskStructArgs
+from sdcm.provision.gce.metadata_provider import MetadataProvider, MetadataArgs
+from sdcm.provision.gce.virtual_machine_provider import VirtualMachineProvider
+from sdcm.provision.provisioner import Provisioner, InstanceDefinition, VmInstance, PricingModel
+from sdcm.utils.gce_utils import get_gce_service
+
+LOGGER = logging.getLogger(__name__)
+
+
+class GCEProvisioner(Provisioner):
+    """Provides API for VM provisioning in GCE"""
+
+    def __init__(self, test_id, region, params: dict):
+        super().__init__(test_id, region)
+        self._params = params
+        self._gce_service = get_gce_service(region=self.region)
+        self._zone = region
+        self._service_accounts = KeyStore().get_gcp_service_accounts()
+        self._ex_disk_struct_provider = DiskStructProvider
+        self._ex_metadata_provider = MetadataProvider
+        self._vm_provider = VirtualMachineProvider(region=region,
+                                                   disk_struct_provider=DiskStructProvider,
+                                                   metadata_provider=MetadataProvider)
+
+    @classmethod
+    def discover_regions(cls, test_id) -> List[Provisioner]:
+        gce_service = get_gce_service(region=GCE_REGIONS.us_east1)
+        gce_locations: list[NodeLocation] = gce_service.list_locations()
+        return [cls(test_id=test_id, region=location.name, params={}) for location in gce_locations]
+
+    @property
+    def cluster_params(self):
+        return self._params
+
+    @cluster_params.setter
+    def cluster_params(self, params: dict):
+        self._params = params
+
+    def get_or_create_instance(self,
+                               definition: InstanceDefinition,
+                               pricing_model: PricingModel = PricingModel.SPOT) -> VmInstance:
+        """
+        Create VmInstance in provided region, specified by InstanceDefinition
+        """
+        gce_project_name = self._gce_service.ex_get_project().name
+
+        disk_struct_args = DiskStructArgs(
```

Comment:
can disk struct_args and metadata_args be created after we verify machine is created (e.g. we fetch it from cache)?